src/providers/virtualraster/qgsvirtualrasterprovider.cpp

```diff
@@ -0,0 +1,338 @@
+
+  if ( !mCalcNode )
+  {
+    mValid = false;
+    return;
+  }
+
+  QStringList rLayerDict = QgsRasterCalcNode::referencedLayerNames( mFormulaString );
+
+  QStringList rasterRefs = QgsRasterCalcNode::cleanRasterReferences( mFormulaString );
+
+  QList<VirtualRasterInputLayers>::iterator it;
+  for ( it = decodedUriParams.rInputLayers.begin(); it != decodedUriParams.rInputLayers.end(); ++it )
+  {
+
+    if ( ! rLayerDict.contains( it->name ) )
+    {
+      mValid = false;
+      return;
+    }
+
+    QgsRasterLayer *rProvidedLayer = new QgsRasterLayer( it->uri, it->name, it->provider );
+
+    if ( ! rProvidedLayer->isValid() )
+    {
+      mValid = false;
+      return;
+    }
+
+    if ( mRasterLayers.contains( rProvidedLayer ) )
+    {
+      continue;
+    }
+
+    //this var is not useful right now except for the copy constructor
+    mRasterLayers << rProvidedLayer;
+
+    for ( int j = 0; j < rProvidedLayer->bandCount(); ++j )
+    {
+      if ( ! rasterRefs.contains( rProvidedLayer->name() + QStringLiteral( "@" ) + QString::number( j + 1 ) ) )
+      {
+        continue;
+      }
+
+      QgsRasterCalculatorEntry entry;
+      entry.raster = rProvidedLayer;
+      entry.bandNumber = j + 1;
+      entry.ref = rProvidedLayer->name() + QStringLiteral( "@" ) + QString::number( j + 1 );
+      mRasterEntries.push_back( entry );
+    }
+  }
+  mValid = true;
+}
+
+QgsVirtualRasterProvider::QgsVirtualRasterProvider( const QgsVirtualRasterProvider &other )
+  : QgsRasterDataProvider( other.dataSourceUri(), QgsDataProvider::ProviderOptions() )
+  , mValid( other.mValid )
+  , mCrs( other.mCrs )
+  , mExtent( other.mExtent )
+  , mWidth( other.mWidth )
+  , mHeight( other.mHeight )
+  , mBandCount( other.mBandCount )
+  , mXBlockSize( other.mXBlockSize )
+  , mYBlockSize( other.mYBlockSize )
+  , mFormulaString( other.mFormulaString )
+  , mLastError( other.mLastError )
+
+{
+  for ( const auto &it : other.mRasterLayers )
+  {
+    QgsRasterLayer *rcProvidedLayer = it->clone();
+    for ( int j = 0; j < rcProvidedLayer->bandCount(); ++j )
+    {
+      QgsRasterCalculatorEntry entry;
+      entry.raster = rcProvidedLayer;
+      entry.bandNumber = j + 1;
+      entry.ref = rcProvidedLayer->name() % QStringLiteral( "@" ) + QString::number( j + 1 );
+      mRasterEntries.push_back( entry );
+    }
+  }
+
+  mCalcNode.reset( QgsRasterCalcNode::parseRasterCalcString( mFormulaString, mLastError ) );
+}
+
+QgsVirtualRasterProvider::~QgsVirtualRasterProvider()
+{
+  qDeleteAll( mRasterLayers );
+}
+
+QgsRasterBlock *QgsVirtualRasterProvider::block( int bandNo, const QgsRectangle &extent, int width, int height, QgsRasterBlockFeedback *feedback )
+{
+  Q_UNUSED( bandNo );
+  std::unique_ptr< QgsRasterBlock > tblock = std::make_unique< QgsRasterBlock >( Qgis::DataType::Float64, width, height );
+  double *outputData = ( double * )( tblock->bits() );
+
+  //else  // Original code (memory inefficient route)
```

Comment:
???

Reply:
I should've delete it, it's from the original qgsrastercalculator class